cpp/include/cudf/fixed_point/floating_conversion.hpp

```diff
@@ -0,0 +1,269 @@
+    return (integer_rep & mantissa_mask) | understood_bit_mask;
+  }
+
+  /** @brief Extracts the sign bit of a floating-point number
+   *
+   * @param floating The floating value to extract the sign from
+   * @return The sign bit
+   */
+  CUDF_HOST_DEVICE inline static bool get_is_negative(FloatingType floating)
+  {
+    // Convert floating to integer
+    auto const integer_rep = bit_cast_to_integer(floating);
+
+    // Extract the sign bit:
+    return static_cast<bool>(sign_mask & integer_rep);
+  }
+
+  /** @brief Extracts the exponent of a floating-point number
+   *
+   * @note This returns INT_MIN for +/-0, +/-inf, NaN's, and denormals
+   * For all of these cases, the decimal fixed_point number should be set to zero
+   *
+   * @param floating The floating value to extract the exponent from
+   * @return The stored base-2 exponent, or INT_MIN for special values
+   */
+  CUDF_HOST_DEVICE inline static int get_exp2(FloatingType floating)
+  {
+    // Convert floating to integer
+    auto const integer_rep = bit_cast_to_integer(floating);
+
+    // First extract the exponent bits and handle its special values.
+    // To minimize branching, all of these special cases will return INT_MIN.
+    // For all of these cases, the decimal fixed_point number should be set to zero.
+    auto const exponent_bits = integer_rep & exponent_mask;
+    if (exponent_bits == 0) {
+      // Because of the understood set-bit not stored in the mantissa, it is not possible
+      // to store the value zero directly. Instead both +/-0 and denormals are represented with
+      // the exponent bits set to zero.
+      // Thus it's fastest to just floor (generally unwanted) denormals to zero.
+      return INT_MIN;
+    } else if (exponent_bits == exponent_mask) {
+      //+/-inf and NaN values are stored with all of the exponent bits set.
+      // As none of these are representable by integers, we'll return the same value for all cases.
+      return INT_MIN;
+    }
+
+    // Extract the exponent value: shift the bits down and subtract the bias.
+    using SignedIntegralType                       = cuda::std::make_signed_t<IntegralType>;
+    SignedIntegralType const shifted_exponent_bits = exponent_bits >> num_mantissa_bits;
+    return shifted_exponent_bits - static_cast<SignedIntegralType>(exponent_bias);
+  }
+
+  /** @brief Sets the sign bit of a floating-point number
+   *
+   * @param floating The floating value to set the sign of
+   * @param is_negative The sign bit to set for the floating-point number
+   * @return The input floating-point value with the chosen sign
+   */
+  CUDF_HOST_DEVICE inline static FloatingType set_is_negative(FloatingType floating,
+                                                              bool is_negative)
+  {
+    // Convert floating to integer
+    IntegralType integer_rep = bit_cast_to_integer(floating);
+
+    // Set the sign bit
+    integer_rep |= (IntegralType(is_negative) << sign_bit_index);
```

Comment:
Is this right? If the sign bit is already `1`, it's not possible to change it with an OR operator, regardless of `is_negative`. Are we assuming the sign bit is always `0` when entering this function? If so, that is not documented.

Are there test cases for this code?

Reply:
The sign is always zero on input, I'll add that comment.  It's not part of this PR, but when I switch from the old decimal/floating conversion code to this, the fixed_point tests test all of this functionality, and all of those tests pass.

Reply:
Comment added

Reply:
Might I suggest a reminder in the comment added that 0 means positive value, so the or will do what we want. Simply saying it must be positive implies the reader knows this.

Reply:
Done

Reply:
I don't like the design of this function.

1. If `is_negative` is true, this should be a no-op -- we shouldn't even cast back and forth.
2. Where is this being called? It doesn't appear to be used in this PR, and it doesn't seem like it's designed for public consumption.
3. Can we avoid bitwise math entirely and just negate the value? i.e. `return is_negative ? -floating : floating`

Reply:
This PR contains utilities for the upcoming primary PR for the decimal <--> floating conversion.  I broke the code into several PRs as it is quite large.  So nothing is calling this yet.  

This specific function is used at the end of decimal --> floating.  In the main algorithm we do a lot of bit-shifting so the sign is initially cleared, and here we are setting it at the end.  Doing it this way (cast + or) requires no branching, so you don't pay a performance penalty for the branch.  A couple months ago I did a bunch of benchmarking trying different methods and this method was the fastest.